src/complete.jl

```diff
@@ -34,11 +35,64 @@ function n_complete(n::Int, D::Int)
     out
 end
 
-@generated function complete_polynomial!{N}(z::Matrix, d::Degree{N},
-                                            out::Matrix)
+#
+# Generating basis functions
+#
+@generated function complete_polynomial!{N,T,Ndim}(z::Array{T,Ndim}, d::Degree{N},
+                                                   out::Array{T,Ndim})
     complete_polynomial_impl!(z, d, out)
 end
 
+function complete_polynomial!{T, Ndim}(z::Array{T,Ndim}, d::Int, out::Array{T,Ndim})
+    complete_polynomial!(z, Degree{d}(), out)::Array{T,Ndim}
+end
+
+
+#
+# Vector versions for generating basis functions
+#
+function complete_polynomial_impl!{T,N}(z::Type{Vector{T}}, ::Type{Degree{N}},
+                                        out::Type{Vector{T}})
+    big_temp = Expr(:(=), Symbol("tmp_$(N+1)"), one(T))
+    quote
+        nvar = length(z)
+        if length(out) != (n_complete(nvar, $N))
+            error("z, out not compatible")
+        end
+
+        # reset first column to ones
+        out[1] = one($T)
+
+        ix = 1
+        $big_temp
+        @nloops($N, # number of loops
+                i,  # counter
+                d->((d == $N ? 1 : i_{d+1}) : nvar),  # ranges
+                d->(d == 1 ?
+                    (begin
+                         ix += 1
+                         out[ix] = tmp_{d+1}*z[i_d]
+                     end) :
+                    (begin
+                        ix += 1
+                        tmp_d = tmp_{d+1}*z[i_d]
+                        out[ix] = tmp_d
+                    end)),  # preexpr
+                Expr(:block, :nothing)  # bodyexpr
+                )
+        out
+    end
+end
+
+function complete_polynomial{T}(z::Vector{T}, d::Int)
+    nvar = length(z)
+    out = Array(T, n_complete(nvar, d))
+    complete_polynomial!(z, Degree{d}(), out)::Vector{T}
```

Comment:
Hmmm. We still get intermediate type instability because it doesn't know what type gets returned by `complete_polynomial!(z, Degree{d}(), out)`... Is that a problem? It really seems like it should be able to infer that type.